### PR TITLE
Managed resource exception handling - fix for #173

### PR DESCRIPTION
**Pull request by argv-minus-one**
This PR adds more complete exception handling for ManagedResource. It now works the same way that Java try-with-resources does:

* All exceptions, including fatal ones like `VirtualMachineError` and `ControlThrowable`, now cause the resource to be disposed of.
* When an exception is thrown while using the resource (we'll call it `e1`), and another is thrown while disposing of the resource (`e2`), the latter is no longer silently ignored. Instead, `e2` is attached as a “[suppressed exception](https://docs.oracle.com/javase/8/docs/api/java/lang/Throwable.html#addSuppressed-java.lang.Throwable-)” to `e1`. As a result, `e1.printStackTrace()` will also show a stack trace for `e2`.

---

You can see the non-local return use-case in `ManagedResourceSpec.scala` starting from line 152. There, the test `return`s from inside a resource-management block, and a managed resource is properly disposed of in the process.

If you run this test on the previous implementation of `ManagedResource` (without the exception handling additions in this PR), the test will fail; the resource will not be disposed of.

This is because the compiler implements such a `return` by throwing a `ControlThrowable` from inside the block, then catching it on the outside. But `ControlThrowable` is considered a fatal exception, so `Try` won't catch it. On the other hand, `catch` and `finally` will.

---

Long story short: this PR should make better-files ARM safer and more predictable.

### CODE REVIEW

**Comment by argv-minus-one** [comment]
>* Avoid using null

This was an intentional decision, to reduce the chance of `OutOfMemoryError` at a time when that is relatively likely to happen.

>* Consider using util.Try instead of a try-catch block
>* Fatal exception is being caught
>* Avoid mutable fields

This is intentional. It is not possible to implement proper ARM exception handling without doing that.

>* Avoid using return

This is an intentional part of the test case.

>* Default case not specified.

False. The pattern `e: Throwable` in a `catch` block matches all possible inputs.

It is possible to say `case e => …` instead, but this generates a Scala compiler warning, and is less readable.

>* [There should be no space before a left bracket ''

False. In the ScalaTest style in use at that position, a space before the bracket is correct. [See ScalaTest user guide.](http://www.scalatest.org/user_guide/using_matchers#expectedExceptions)

>* Public method must have explicit type

It's a public method *on a private class.* Still, I'll go ahead and add explicit return types.

>* File line length exceeds 160 characters
>* Use braces in for comprehensions
>* The string literal "no exception was thrown" appears 2 times in the file.

This is silly, but very well, I'll add a commit to resolve these issues.

>* The string literal "one" appears 7 times in the file.
>* The string literal "hello" appears 4 times in the file.

The test code would be *less* readable if I “fixed” that.

**Comment by pathikrit** [comment]
Sorry ignore the codacy-bot.
